src/common/eeprom.c

```diff
@@ -2,230 +2,295 @@
-    VARIANT8_UI8,
-    VARIANT8_UI8,
-    VARIANT8_UI8,
-    VARIANT8_UI8,
-    VARIANT8_UI8,
-    VARIANT8_UI8,
-    VARIANT8_UI8,
-    VARIANT8_UI8,
-    VARIANT8_UI8,
+#include "cmsis_os.h"
+#include "ff.h"
+
+
+
+#define EEPROM_VARCOUNT  (sizeof(eeprom_map) / sizeof(eeprom_entry_t))
+#define EEPROM_DATASIZE  sizeof(eeprom_vars_t)
+#define EEPROM__PADDING  1
+
+#pragma pack(push)
+#pragma pack(1)
+
+// eeprom map entry structure
+typedef struct _eeprom_entry_t {
+    uint8_t type;   // variant8 data type
+    uint8_t count;  // number of elements
+} eeprom_entry_t;
+
+// eeprom vars structure (used for defaults)
+typedef struct _eeprom_vars_t {
+    uint16_t VERSION;
+    uint16_t FEATURES;
+    uint16_t DATASIZE;
+    uint16_t FWVERSION;
+    uint16_t FWBUILD;
+    uint8_t FILAMENT_TYPE;
+    uint32_t FILAMENT_COLOR;
+    uint8_t RUN_SELFTEST;
+    uint8_t RUN_XYZCALIB;
+    uint8_t RUN_FIRSTLAY;
+    uint8_t FSENSOR_ENABLED;
+    float ZOFFSET;
+    float PID_NOZ_P;
+    float PID_NOZ_I;
+    float PID_NOZ_D;
+    float PID_BED_P;
+    float PID_BED_I;
+    float PID_BED_D;
+    uint8_t LAN_FLAG;
+    uint32_t LAN_IP4_ADDR;
+    uint32_t LAN_IP4_MSK;
+    uint32_t LAN_IP4_GW;
+    uint32_t LAN_IP4_DNS1;
+    uint32_t LAN_IP4_DNS2;
+    char LAN_HOSTNAME[LAN_HOSTNAME_MAX_LEN + 1];
+    char TEST[10];
+    char _PADDING[EEPROM__PADDING];
+    uint32_t CRC32;
+} eeprom_vars_t;
+
+#pragma pack(pop)
+
+
+// eeprom map
+const eeprom_entry_t eeprom_map[] = {
+    { VARIANT8_UI16,  1  }, // EEVAR_VERSION
+    { VARIANT8_UI16,  1  }, // EEVAR_FEATURES
+    { VARIANT8_UI16,  1  }, // EEVAR_DATASIZE
+    { VARIANT8_UI16,  1  }, // EEVAR_FW_VERSION
+    { VARIANT8_UI16,  1  }, // EEVAR_FW_BUILD
+    { VARIANT8_UI8,   1  }, // EEVAR_FILAMENT_TYPE
+    { VARIANT8_UI32,  1  }, // EEVAR_FILAMENT_COLOR
+    { VARIANT8_UI8,   1  }, // EEVAR_RUN_SELFTEST
+    { VARIANT8_UI8,   1  }, // EEVAR_RUN_XYZCALIB
+    { VARIANT8_UI8,   1  }, // EEVAR_RUN_FIRSTLAY
+    { VARIANT8_UI8,   1  }, // EEVAR_FSENSOR_ENABLED
+    { VARIANT8_FLT,   1  }, // EEVAR_ZOFFSET
+    { VARIANT8_FLT,   1  }, // EEVAR_PID_NOZ_P
+    { VARIANT8_FLT,   1  }, // EEVAR_PID_NOZ_I
+    { VARIANT8_FLT,   1  }, // EEVAR_PID_NOZ_D
+    { VARIANT8_FLT,   1  }, // EEVAR_PID_BED_P
+    { VARIANT8_FLT,   1  }, // EEVAR_PID_BED_I
+    { VARIANT8_FLT,   1  }, // EEVAR_PID_BED_D
+    { VARIANT8_UI8,   1  }, // EEVAR_LAN_FLAG
+    { VARIANT8_UI32,  1  }, // EEVAR_LAN_IP4_ADDR
+    { VARIANT8_UI32,  1  }, // EEVAR_LAN_IP4_MSK
+    { VARIANT8_UI32,  1  }, // EEVAR_LAN_IP4_GW
+    { VARIANT8_UI32,  1  }, // EEVAR_LAN_IP4_DNS1
+    { VARIANT8_UI32,  1  }, // EEVAR_LAN_IP4_DNS2
+    { VARIANT8_PCHAR, LAN_HOSTNAME_MAX_LEN + 1 }, // EEVAR_LAN_HOSTNAME
+    { VARIANT8_PCHAR, 10 }, // EEVAR_TEST
+    { VARIANT8_PCHAR, EEPROM__PADDING }, // EEVAR__PADDING32
+    { VARIANT8_UI32,  1  }, // EEVAR_CRC32
 };
 
+// eeprom variable names
 const char *eeprom_var_name[] = {
     "VERSION",
+    "FEATURES",
```

Comment:
Keeping all these descriptor arrays in sync will be very hard to maintain.
I don't see any compile-time checks to prevent accidental errors.
Probably making a:
struct EEPROMItemDesc {
  char var_name[14];
  uint8_t type;
  uint8_t bytes_used;
  variantXX default_value;
};
would force users add all necessary components when adding a new eeprom var

Reply:
Names moved to eeprom_tab structure (commit 0ef1558), in tis time definition of eeprom data structure has two parts - first is eeprom_tab (used to access vars) and second is structure with default values (used in reset defaults and import previous version).